Lib/test/test_pty.py

```diff
@@ -262,8 +350,9 @@ def _socketpair(self):
         self.files.extend(socketpair)
         return socketpair
 
-    def _mock_select(self, rfds, wfds, xfds):
+    def _mock_select(self, rfds, wfds, xfds, timeout=0):
```

Comment:
Yes! Glad that you asked! This is actually an interesting story. `SmallPtyTests.test__copy_eof_on_all()` emulates a situation where both `stdin` and `master` reach EOF in the same iteration of `pty.spawn()`'s copy loop ( `_copy()` in Lib/pty.py ). This leads to both of them being removed from `select()`'s fd list at once. While this does not occur often in practical scenarios on Linux, FreeBSD, etc, this is a situation that must be considered [ maybe some other platform behaves this way/some special case ]; that is why this test exists in the first place! Now, at least on Linux, select.select([],[],[]) will block indefinitely [ all three args are empty lists because `master` and `stdin` were kicked out ]. In such situation, one should set a timeout for `select.select()` once both `master` and `stdin` are out. After the timeout is reached, we exit from the loop gracefully. Now, one might ask why not just put a `while fds:` there instead of the `while True` to avoid setting a timeout. The reason is that `SmallPtyTests.test__copy_eof_on_all()` EXPECTS the copy loop to make ONE ADDITIONAL iteration of the loop before it exits. In fact, this is how I have seen it being done in other programs; the util-linux version of script(1) sets a timeout for `poll()` instead of abruptly exiting from its main loop when no more data is available. Therefore, pypty2 also does it this way, so that `SmallPtyTests.test__copy_eof_on_all()` works as expected without modification; only `_mock_select()` must now accept a timeout arg to be compatible; anyway, it ignores the timeout :D whereas the actual `select.select()` does not.

Reply:
Update: I am willing to remove the timeout hack in `pty._copy()` and replace it with the cleaner `while fds:` alternative that I mentioned. However, that will require modifying/removing `SmallPtyTests.test__copy_eof_on_all()`, which I should only do after I have convinced the developers that everything else works correctly :)